### PR TITLE
docs: Explain how FlowStages work and their rationale

### LINKED ISSUE #557: Document how Test Suite FlowStages work, including the philosophy behind them
Use this PR as inspiration: https://github.com/openactive/openactive-test-suite/pull/552

Include how to read each flow, and what input and outputs are about.

### PR DESCRIPTION
**Pull request by lukehesluke**
Closes #557

### CODE REVIEW

File: packages/openactive-integration-tests/test/helpers/flow-stages/README.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# Flow Stages
+
+A Flow is a sequence of API calls that are made by Test Suite, for a given test, against either the Booking System  or the Broker Microservice. A Flow Stage encapsulates the configuration and logic for one of those API calls. e.g. there is a FlowStage for C1 (which is in the Booking System), a FlowStage for Fetch Opportunities (which uses the Broker Microservice), etc.
```
Comment (civsiv) [inline packages/openactive-integration-tests/test/helpers/flow-stages/README.md:3]:
```suggestion
A Flow is a sequence of API calls that are made by Test Suite, for a given test, against either the Booking System  or the Broker Microservice. A Flow is composed of several Flow Stages. A Flow Stage encapsulates the configuration and logic for one of those API calls. e.g. there is a FlowStage for C1 (which is in the Booking System), a FlowStage for Fetch Opportunities (which uses the Broker Microservice), etc.
```

File: packages/openactive-integration-tests/test/helpers/flow-stages/README.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  - Using `runFn(..)`, which is provided in the constructor
+3. Transforms the output
+  - This is also performed in `runFn(..)`, which is provided in the constructor
+4. Sends (some of) this output to subsequent FlowStages
+  - This output is automatically saved by the FlowStage as the output from `runFn(..)`, which is provided in the constructor, and can be retrieved by calling `getOutput(..)` on the FlowStage instance.
+
+FlowStages can then be queried after they've run in order to:
+
+1. Check that the FlowStage was successful
+  - This is overridable per-FlowStage. For example, B's Flow Stage considers the run successful if the HTTP response has status 201
```
Comment (civsiv) [inline packages/openactive-integration-tests/test/helpers/flow-stages/README.md:38]:
```suggestion
  - This is overridable per-FlowStage. For example, B's Flow Stage considers the run successful if the HTTP response has status 201 as opposed to the default FlowStage behaviour that checks for status 200
```

File: packages/openactive-integration-tests/test/helpers/flow-stages/README.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  - FlowStage method: `itSuccessChecks()`
+2. Perform validation checks on the output
+  - This is overridable per-FlowStage. In all cases, this is a case of calling [Validator](https://github.com/openactive/data-model-validator) on the HTTP output.
```
Comment (civsiv) [inline packages/openactive-integration-tests/test/helpers/flow-stages/README.md:41]: Same comment here re overridable

File: packages/openactive-integration-tests/test/helpers/flow-stages/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+An abstraction that can be either a **Flow Stage**, [**Book Recipe**](./book-recipe.js) or a [**Flow Stage Run**](./flow-stage-run.js). This represents a single FlowStage or sequence of FlowStages that can be run.
+
+This encapsulation allows us to, for example, more easily reason about tests in which the "book" stage could use either [Simple Booking Flow](https://openactive.io/open-booking-api/EditorsDraft/#simple-booking-flow) or [Booking Flow with Approval](https://openactive.io/open-booking-api/EditorsDraft/#booking-flow-with-approval), which involve different sets of API calls.
+
+## Jest Tests
+
+Flows, consisting of Flow Stages, run the underlying API calls which are being tested, via Jest in the various Test Suite tests.
+
+Jest tests involve a custom course of execution, in which setup occurs in `beforeEach`/`beforeAll` hooks, tests are run in `it` hooks, etc. Flow Stages are designed to work with this.
```
Comment (civsiv) [inline packages/openactive-integration-tests/test/helpers/flow-stages/README.md:54]:
```suggestion
Jest tests involve a custom course of execution, in which setup occurs in `beforeEach`/`beforeAll` hooks and tests are run in `it` hooks, etc. Flow Stages are designed to work with this.
```

File: packages/openactive-integration-tests/test/helpers/flow-stages/README.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+## FlowStageUtils
+
+[FlowStageUtils](./flow-stage-utils.js) is a collection of utility functions which simplify and standardise the process of using FlowStages to write Test Suite tests.
+
+Of particular importance is the `describeRunAndRunChecks(..)` function, which creates a Jest `describe(..)` hook for a given **FlowStageRunnable** which performs the running and checking of the runnable as described in the **Jest Tests** section.
+
+## FlowStageRecipes
+
+Setting up a flow, consisting of multiple FlowStages where each feeds input into the next, can require a lot of boilerplate and repeated logic. In most cases, parts of these flows can be packaged up as they will be the same in lots of different tests.
```
Comment (civsiv) [inline packages/openactive-integration-tests/test/helpers/flow-stages/README.md:71]:
```suggestion
Setting up a Flow, consisting of multiple FlowStages where each feeds input into the next, can require a lot of boilerplate and repeated logic. In most cases, parts of these flows can be packaged up as they will be the same in lots of different tests.
```

**Review by civsiv — CHANGES_REQUESTED** [review]

File: packages/openactive-integration-tests/test/helpers/flow-stages/README.md (hunk 2 above)
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/helpers/flow-stages/README.md:38]: DONE

File: packages/openactive-integration-tests/test/helpers/flow-stages/README.md (hunk 4 above)
Comment (lukehesluke) [reply packages/openactive-integration-tests/test/helpers/flow-stages/README.md]: i've done it slightly differently involving semicolons

**Comment by lukehesluke** [comment]
@civsiv  confirmed this as approved on another channel

**Pull request merged**